### PR TITLE
[E-Documents Core] - Manual multiple files import

### PR DESCRIPTION
**Pull request by petemchlk**
### Manual multiple files import

This pull request contains changes that make possible to upload multiple files at once if user choose to create E-Documents using manual Import Files action on E-Documents page.

- New fileuploadaction "Import Files" added to E-Documents page (6122)
- Action "New From File" hidden as its functionality is covered by new action
- Changes in "E-Doc. Import" codeunit to handle import using FileUpload list

MS PR: https://github.com/microsoft/ALAppExtensions/pull/27975

### CODE REVIEW

File: Apps/W1/EDocument/app/src/Document/EDocuments.Page.al [hunk 1]
```diff
@@ -65,12 +65,29 @@ page 6122 "E-Documents"
                 Caption = 'New From File';
                 ToolTip = 'Create an electronic document by manually uploading a file.';
                 Image = Import;
+                Visible = false;
```
Comment (GMatuleviciute) [inline Apps/W1/EDocument/app/src/Document/EDocuments.Page.al:68]: I would also add preprocessing symbols #if not CLEAN26

File: Apps/W1/EDocument/app/src/Document/EDocuments.Page.al [hunk 2]
```diff
@@ -65,12 +65,29 @@ page 6122 "E-Documents"
...
                     NewFromFile();
                 end;
             }
+            fileuploadaction(ImportManuallyMultiple)
+            {
+                Caption = 'Import Files';
+                ToolTip = 'Create electronic documents by uploading single or multiple files.';
+                Image = Import;
+                AllowedFileExtensions = '.xml';
+                AllowMultipleFiles = true;
```
Comment (GMatuleviciute) [inline Apps/W1/EDocument/app/src/Document/EDocuments.Page.al:81]: The empty line should be between properties and trigger

File: Apps/W1/EDocument/app/src/Document/EDocuments.Page.al [hunk 3]
```diff
@@ -65,12 +65,29 @@ page 6122 "E-Documents"
...
+                trigger OnAction(Files: List of [FileUpload])
+                var
+                    EDocImport: Codeunit "E-Doc. Import";
+                begin
+                    EDocImport.UploadDocuments(Files);
+                    CurrPage.Update();
```
Comment (GMatuleviciute) [inline Apps/W1/EDocument/app/src/Document/EDocuments.Page.al:87]: Add explicit RunTrigger. Very recommend using false if that's enough for the solution

File: Apps/W1/EDocument/app/src/Document/EDocuments.Page.al [hunk 4]
```diff
@@ -65,12 +65,29 @@ page 6122 "E-Documents"
...
+                end;
+
```
Comment (GMatuleviciute) [inline Apps/W1/EDocument/app/src/Document/EDocuments.Page.al:89]: Unneeded empty line

File: Apps/W1/EDocument/app/src/Integration/EDocIntegrationManagement.Codeunit.al [hunk 5]
```diff
@@ -235,6 +239,24 @@ codeunit 6134 "E-Doc. Integration Management"
     end;
```
Comment (GMatuleviciute) [inline Apps/W1/EDocument/app/src/Integration/EDocIntegrationManagement.Codeunit.al:240]: Two empty lines between procedures, one should be only

File: Apps/W1/EDocument/app/src/Integration/EDocIntegrationManagement.Codeunit.al [hunk 6]
```diff
@@ -235,6 +239,24 @@ codeunit 6134 "E-Doc. Integration Management"
...
 
 
+    internal procedure HasDuplicate(var IncomingEDocument: Record "E-Document"; var EDocumentContent: Codeunit "Temp Blob"; IEDocument: Interface "E-Document"): Boolean
```
Comment (GMatuleviciute) [inline Apps/W1/EDocument/app/src/Integration/EDocIntegrationManagement.Codeunit.al:242]: IEDocument?  Sounds strange and does not really have a reference about what variable is

File: Apps/W1/EDocument/app/src/Integration/EDocIntegrationManagement.Codeunit.al [hunk 7]
```diff
@@ -235,6 +239,24 @@ codeunit 6134 "E-Doc. Integration Management"
...
+    var
+        EDocument: Record "E-Document";
+        EDocGetBasicInfo: Codeunit "E-Doc. Get Basic Info";
+    begin
+        // Commit before getting basic info with error handling (if Codeunit.Run then )
```
Comment (GMatuleviciute) [inline Apps/W1/EDocument/app/src/Integration/EDocIntegrationManagement.Codeunit.al:247]: // Commit before getting basic info with error handling (if Codeunit.Run then)

File: Apps/W1/EDocument/app/src/Processing/EDocImport.Codeunit.al [hunk 8]
```diff
@@ -19,33 +20,38 @@ codeunit 6140 "E-Doc. Import"
...
-        end else begin
-            EDocument.Modify(true);
-            EDocumentProcessing.ModifyServiceStatus(EDocument, EDocumentService, Enum::"E-Document Service Status"::Imported);
-        end;
+    internal procedure UploadDocuments(Documents: List of [FileUpload])
+    var
+        EDocument: Record "E-Document";
+        EDocumentService: Record "E-Document Service";
+    begin
+        if Documents.Count = 0 then
```
Comment (GMatuleviciute) [inline Apps/W1/EDocument/app/src/Processing/EDocImport.Codeunit.al:48]: if Documents.Count() = 0 then

File: Apps/W1/EDocument/app/src/Processing/EDocImport.Codeunit.al [hunk 9]
```diff
@@ -303,7 +309,8 @@ codeunit 6140 "E-Doc. Import"
         UpdateEDocumentRecordId(EDocument, EDocument."Document Type", DocNo, RecordId);
     end;
 
-    local procedure UpdateEDocumentRecordId(var EDocument: Record "E-Document"; EDocType: enum "E-Document Type"; DocNo: Code[20]; RecordId: RecordId)
+    local procedure UpdateEDocumentRecordId(var EDocument: Record "E-Document"; EDocType: enum "E-Document Type"; DocNo: Code[20];
+                                                                                              RecordId: RecordId)
```
Comment (GMatuleviciute) [inline Apps/W1/EDocument/app/src/Processing/EDocImport.Codeunit.al:313]: I think this change is here by accident. Please revert it

File: Apps/W1/EDocument/app/src/Processing/EDocImport.Codeunit.al [hunk 10]
```diff
@@ -651,6 +755,10 @@ codeunit 6140 "E-Doc. Import"
         DocTypeIsNotSupportedErr: Label 'Document type %1 is not supported.', Comment = '%1 - Document Type';
         FailedToFindVendorErr: Label 'No vendor is set for Edocument';
         CannotProcessEDocumentMsg: Label 'Cannot process E-Document %1 with Purchase Order %2 before Purchase Order has been matched and posted for E-Document %3.', Comment = '%1 - E-Document entry no, %2 - Purchase Order number, %3 - EDocument entry no.';
+        DocNotCreatedQst: Label 'Failed to create new %1 from E-Document. Do you want to open E-Document and see the reported errors?', Comment = '%1 - E-Document Document Type';
+        EDocumentAlreadyExistErr: Label 'E-Document with %1 %2, %3 %4 and %5 %6 already exists.', Comment = '%1 - Incoming E-Document No. fieldcaption, %2 - Incoming E-Document No. value, %3 - Bill-to/Pay-to No. fieldcaption, %4 - Bill-to/Pay-to No. value, %5 - Document Date fieldcaption, %6 - Document Date value.';
```
Comment (GMatuleviciute) [inline Apps/W1/EDocument/app/src/Processing/EDocImport.Codeunit.al:759]: EDocumentAlreadyExistErr: Label 'E-Document with %1 %2, %3 %4 and %5 %6 already exists.', Comment = '%1 - Incoming E-Document No. field caption, %2 - Incoming E-Document No. value, %3 - Bill-to/Pay-to No. field caption, %4 - Bill-to/Pay-to No. value, %5 - Document Date field caption, %6 - Document Date value.';

File: Apps/W1/EDocument/test/src/Receive/EDocReceiveManualTest.Codeunit.al [hunk 11]
```diff
@@ -0,0 +1,267 @@
+codeunit 139501 "E-Doc. Manual Import Test"
```
Comment (GMatuleviciute) [inline Apps/W1/EDocument/test/src/Receive/EDocReceiveManualTest.Codeunit.al:1]: Have you checked if the changes do not break any of the existing tests?

File: Apps/W1/EDocument/test/src/Receive/EDocReceiveManualTest.Codeunit.al [hunk 12]
```diff
@@ -0,0 +1,267 @@
...
+        IncommingDocNo: Text[20];
+        DocInstream: InStream;
+    begin
+        TempXMLBuffer.LoadFromText(EDocReceiveFiles.GetDocument1());
+        TempXMLBuffer.Reset();
+        TempXMLBuffer.SetRange(Type, TempXMLBuffer.Type::Element);
+        TempXMLBuffer.SetRange(Path, '/Invoice/cac:AccountingSupplierParty/cac:Party/cbc:EndpointID');
+        TempXMLBuffer.FindFirst();
+        TempXMLBuffer.Value := DocumentVendor."VAT Registration No.";
+        TempXMLBuffer.Modify();
```
Comment (GMatuleviciute) [inline Apps/W1/EDocument/test/src/Receive/EDocReceiveManualTest.Codeunit.al:128]: add explicit run trigger

File: Apps/W1/EDocument/test/src/Receive/EDocReceiveManualTest.Codeunit.al [hunk 13]
```diff
@@ -0,0 +1,267 @@
...
+
+        TempXMLBuffer.Reset();
+        TempXMLBuffer.SetRange(Path, '/Invoice/cbc:ID');
+        TempXMLBuffer.FindFirst();
+        IncommingDocNo := LibraryRandom.RandText(20);
+        TempXMLBuffer.Value := IncommingDocNo;
+        TempXMLBuffer.Modify();
```
Comment (GMatuleviciute) [inline Apps/W1/EDocument/test/src/Receive/EDocReceiveManualTest.Codeunit.al:135]: add explicit run trigger

File: Apps/W1/EDocument/app/src/Document/EDocuments.Page.al (hunk 1 above)
Comment (tinestaric) [reply Apps/W1/EDocument/app/src/Document/EDocuments.Page.al:68]: If it should be preprocessed, then it should also be obsoleted...

File: Apps/W1/EDocument/app/src/Document/EDocuments.Page.al [hunk 14]
```diff
@@ -88,7 +105,8 @@ page 6122 "E-Documents"
         }
         area(Promoted)
         {
-            actionref(Promoted_ImportManually; ImportManually) { }
+            actionref(Promoted_ImportManually; ImportManually) { Visible = false; }
```
Comment (tinestaric) [inline Apps/W1/EDocument/app/src/Document/EDocuments.Page.al:108]: and then this should be obsoleted and preprocessed as well...

File: Apps/W1/EDocument/app/src/Integration/EDocIntegrationManagement.Codeunit.al (hunk 6 above)
Comment (tinestaric) [reply Apps/W1/EDocument/app/src/Integration/EDocIntegrationManagement.Codeunit.al:242]: I'd blame the name of the interface for that. I wouldn't be too much against the parameter name.

File: Apps/W1/EDocument/app/src/Processing/EDocImport.Codeunit.al [hunk 15]
```diff
@@ -620,6 +627,39 @@ codeunit 6140 "E-Doc. Import"
             until TempEDocImportedLine.Next() = 0;
     end;
 
+    local procedure ChooseEDocumentService(var EDocumentService: Record "E-Document Service"): Boolean
+    begin
+        exit(not (Page.RunModal(Page::"E-Document Services", EDocumentService) <> Action::LookupOK));
```
Comment (tinestaric) [inline Apps/W1/EDocument/app/src/Processing/EDocImport.Codeunit.al:632]: wouldn't this be simplified with exit(Page.RunModal() = LookUpOk)?

File: Apps/W1/EDocument/app/src/Processing/EDocImport.Codeunit.al [hunk 16]
```diff
@@ -19,33 +20,38 @@ codeunit 6140 "E-Doc. Import"
...
+    end;
 
-        if EDocument."Entry No" = 0 then begin
-            EDocument.Insert(true);
-            EDocumentProcessing.InsertServiceStatus(EDocument, EDocumentService, Enum::"E-Document Service Status"::Imported);
-        end else begin
-            EDocument.Modify(true);
-            EDocumentProcessing.ModifyServiceStatus(EDocument, EDocumentService, Enum::"E-Document Service Status"::Imported);
-        end;
+    internal procedure UploadDocuments(Documents: List of [FileUpload])
```
Comment (tinestaric) [inline Apps/W1/EDocument/app/src/Processing/EDocImport.Codeunit.al:43]: not a fan of coupling UI elements to business logic. 
It would be better to handle the "ChooseEDocumentService" on the page. Select the service, and if you do, then call UploadDocuments(Documents, Service)

File: Apps/W1/EDocument/app/src/Document/EDocuments.Page.al (hunk 1 above)
Comment (petemchlk) [reply Apps/W1/EDocument/app/src/Document/EDocuments.Page.al:68]: thought maybe its better if they add it together with obsolete tags

File: Apps/W1/EDocument/app/src/Integration/EDocIntegrationManagement.Codeunit.al (hunk 6 above)
Comment (petemchlk) [reply Apps/W1/EDocument/app/src/Integration/EDocIntegrationManagement.Codeunit.al:242]: **I**EDocument meaning its interface. Procedure was merged from other pull request. We can change that to InterfaceEDocument or EDocumentInterface but then I we should also change it probably in other PR

File: Apps/W1/EDocument/app/src/Integration/EDocIntegrationManagement.Codeunit.al (hunk 7 above)
Comment (petemchlk) [reply Apps/W1/EDocument/app/src/Integration/EDocIntegrationManagement.Codeunit.al:247]: Should remove this? Wouldn't it be better to have comment on why we do the commit? They have the same comment when they call GetBasicInfo with commit before

File: Apps/W1/EDocument/app/src/Processing/EDocImport.Codeunit.al (hunk 9 above)
Comment (petemchlk) [reply Apps/W1/EDocument/app/src/Processing/EDocImport.Codeunit.al:313]: yes it is. Thanks for catching this

File: Apps/W1/EDocument/app/src/Processing/EDocImport.Codeunit.al (hunk 10 above)
Comment (petemchlk) [reply Apps/W1/EDocument/app/src/Processing/EDocImport.Codeunit.al:759]: Should this message be less complicated?

File: Apps/W1/EDocument/app/src/Processing/EDocImport.Codeunit.al (hunk 16 above)
Comment (tinestaric) [reply Apps/W1/EDocument/app/src/Processing/EDocImport.Codeunit.al:43]: ahh, I see you only abstracted the above one...

Comment (petemchlk) [reply Apps/W1/EDocument/app/src/Processing/EDocImport.Codeunit.al:43]: yes, but it can be fixed since I'm touching it anyways :D

File: Apps/W1/EDocument/app/src/Processing/EDocImport.Codeunit.al [hunk 17]
```diff
@@ -19,33 +20,38 @@ codeunit 6140 "E-Doc. Import"
...
             exit;
 
-        TempBlob.CreateOutStream(OutStr);
-        CopyStream(OutStr, InStr);
+        this.ImportEDocumentFromStream(EDocument, EDocumentService, InStr);
+    end;
 
-        EDocument.Direction := EDocument.Direction::Incoming;
-        EDocument."Document Type" := Enum::"E-Document Type"::None;
+    internal procedure UploadDocument(var EDocument: Record "E-Document"; EDocumentService: Record "E-Document Service"; DocumentFile: FileUpload): Boolean
```
Comment (tinestaric) [inline Apps/W1/EDocument/app/src/Processing/EDocImport.Codeunit.al:35]: Where is this one used?

**Review by GMatuleviciute — CHANGES_REQUESTED** [review]

File: Apps/W1/EDocument/app/src/Processing/EDocImport.Codeunit.al [hunk 18]
```diff
@@ -637,6 +677,70 @@ codeunit 6140 "E-Doc. Import"
     end;
 #endif
 
+    local procedure HandleMultipleDocumentUpload(var Documents: List of [FileUpload]; var EDocument: Record "E-Document"; EDocumentService: Record "E-Document Service")
```
Comment (tinestaric) [inline Apps/W1/EDocument/app/src/Processing/EDocImport.Codeunit.al:680]: This feels like a very hacky approach. Maybe we could have a call around this instead of comments?

**Review by tinestaric — CHANGES_REQUESTED** [review]

File: Apps/W1/EDocument/app/src/Processing/EDocImport.Codeunit.al [hunk 19]
```diff
@@ -637,6 +657,83 @@ codeunit 6140 "E-Doc. Import"
...
+            end;
+        end;
+
+        if NotProcessedDocuments > 0 then
+            Message(StrSubstNo(DuplicatesMsg, NotProcessedDocuments, Documents.Count()))
+        else
+            Message(DocsImportedMsg);
+    end;
+
+    internal procedure HandleSingleDocumentUpload(Document: FileUpload; EDocument: Record "E-Document"; EDocumentService: Record "E-Document Service")
```
Comment (tinestaric) [inline Apps/W1/EDocument/app/src/Processing/EDocImport.Codeunit.al:691]: To make this procedure testable, you could have the InStream parameter instead of FileUpload?
Documents.Get(1).CreateInStream(InStr) and then pass just the stream? Then you can test the duplicate error, the confirmation message, and page run

File: Apps/W1/EDocument/app/src/Processing/EDocImport.Codeunit.al [hunk 20]
```diff
@@ -637,6 +657,83 @@ codeunit 6140 "E-Doc. Import"
...
+    var
+        TempBlob: Codeunit "Temp Blob";
+        DocumentInstream: InStream;
+        DocumentOutStream: OutStream;
+    begin
+        TempBlob.CreateOutStream(DocumentOutStream);
```
Comment (tinestaric) [inline Apps/W1/EDocument/app/src/Processing/EDocImport.Codeunit.al:697]: Something I recently discovered is that CreateOutStream can also return the OutStream without passing it as a parameter. 

So you can do things like 
CopyStream(TempBlob.CreateOutStream(), DocumentInStream);

File: Apps/W1/EDocument/app/src/Processing/EDocImport.Codeunit.al [hunk 21]
```diff
@@ -637,6 +657,83 @@ codeunit 6140 "E-Doc. Import"
...
+        if not this.HideDialogs and EDocErrorHelper.HasErrors(EDocument) then
+            if Confirm(DocNotCreatedQst, true, EDocument."Document Type") then
+                Page.Run(Page::"E-Document", EDocument);
+    end;
+
+    local procedure HasDuplciate(var EDocument: Record "E-Document"; var TempBlob: Codeunit "Temp Blob"; IEDocument: Interface "E-Document"): Boolean
+    var
+        IntegrationManagement: Codeunit "E-Doc. Integration Management";
+    begin
+        if IntegrationManagement.HasDuplicate(EDocument, TempBlob, IEDocument) then
```
Comment (tinestaric) [inline Apps/W1/EDocument/app/src/Processing/EDocImport.Codeunit.al:723]: this could just exit the return value directly?

File: Apps/W1/EDocument/app/src/Processing/EDocImport.Codeunit.al [hunk 22]
```diff
@@ -637,6 +657,83 @@ codeunit 6140 "E-Doc. Import"
...
+        DocumentOutStream: OutStream;
+        NotProcessedDocuments: Integer;
+    begin
+        this.SetHideDialogs(true);
+
+        foreach Document in Documents do begin
+            Clear(EDocument);
+            Clear(DocumentInstream);
+            Clear(TempBlob);
+            TempBlob.CreateOutStream(DocumentOutStream);
```
Comment (tinestaric) [inline Apps/W1/EDocument/app/src/Processing/EDocImport.Codeunit.al:674]: same CreateOutStream remark here

File: Apps/W1/EDocument/app/src/Processing/EDocImport.Codeunit.al [hunk 23]
```diff
@@ -637,6 +657,83 @@ codeunit 6140 "E-Doc. Import"
...
+            Document.CreateInStream(DocumentInstream);
+            CopyStream(DocumentOutStream, DocumentInstream);
+            if HasDuplciate(EDocument, TempBlob, EDocumentService."Document Format") then
+                NotProcessedDocuments += 1
+            else begin
+                TempBlob.CreateInStream(DocumentInstream); // Restore InSteram as it was consumed by CopyStream
```
Comment (tinestaric) [inline Apps/W1/EDocument/app/src/Processing/EDocImport.Codeunit.al:680]: You can now do that with InStream.ResetPosition();

File: Apps/W1/EDocument/app/src/Processing/EDocImport.Codeunit.al [hunk 24]
```diff
@@ -620,6 +614,34 @@ codeunit 6140 "E-Doc. Import"
...
 
+    internal procedure ChooseEDocumentService(var EDocumentService: Record "E-Document Service"): Boolean
+    begin
+        exit(Page.RunModal(Page::"E-Document Services", EDocumentService) = Action::LookupOK);
+    end;
+
+    local procedure ImportEDocumentFromStream(var EDocument: Record "E-Document"; EDocumentService: Record "E-Document Service"; var InStr: InStream)
+    var
+        TempBlob: Codeunit "Temp Blob";
+        OutStr: OutStream;
```
Comment (tinestaric) [inline Apps/W1/EDocument/app/src/Processing/EDocImport.Codeunit.al:625]: redundant outStream var

File: Apps/W1/EDocument/app/src/Integration/EDocIntegrationManagement.Codeunit.al [hunk 25]
```diff
@@ -235,6 +239,24 @@ codeunit 6134 "E-Doc. Integration Management"
     end;
 
 
+    internal procedure HasDuplicate(var IncomingEDocument: Record "E-Document"; var EDocumentContent: Codeunit "Temp Blob"; IEDocument: Interface "E-Document"): Boolean
+    var
+        EDocument: Record "E-Document";
+        EDocGetBasicInfo: Codeunit "E-Doc. Get Basic Info";
+    begin
+        // Commit before getting basic info with error handling (if Codeunit.Run then)
+        Commit();
```
Comment (tinestaric) [inline Apps/W1/EDocument/app/src/Integration/EDocIntegrationManagement.Codeunit.al:248]: I know this is from another PR, but this commit in here makes me a bit nervous. Because seeing a procedure like HasDuplicates doesn't make you think that the transaction will be commited in here...

You don't have to change anything, because I know you have other fires to fight, and I know MS is quite liberal with their commits as well, but just something to think about...

File: Apps/W1/EDocument/app/src/Processing/EDocImport.Codeunit.al [hunk 26]
```diff
@@ -637,6 +659,75 @@ codeunit 6140 "E-Doc. Import"
...
+        foreach Document in Documents do begin
+            Clear(EDocument);
+            Clear(DocumentInstream);
+            Clear(TempBlob);
+            Document.CreateInStream(DocumentInstream);
+            CopyStream(TempBlob.CreateOutStream(), DocumentInstream);
+            if HasDuplciate(EDocument, TempBlob, EDocumentService."Document Format") then
+                NotProcessedDocuments += 1
+            else begin
+                DocumentInstream.ResetPosition();
```
Comment (tinestaric) [inline Apps/W1/EDocument/app/src/Processing/EDocImport.Codeunit.al:680]: very much a nit: but the reset position could be inside the CreateEdocumentFromStream.

File: Apps/W1/EDocument/test/src/Receive/EDocManualImportTest.Codeunit.al [hunk 27]
```diff
@@ -0,0 +1,198 @@
...
+        LibraryEDoc: Codeunit "Library - E-Document";
+        LibraryPurchase: Codeunit "Library - Purchase";
+        LibraryVariableStorage: Codeunit "Library - Variable Storage";
+        EDocImplState: Codeunit "E-Doc. Impl. State";
+        Assert: Codeunit Assert;
+        EDocReceiveFiles: Codeunit "E-Doc. Receive Files";
+        LibraryRandom: Codeunit "Library - Random";
+
+    [Test]
+    procedure ManuallyCreateEDocumentFromStream()
```
Comment (tinestaric) [inline Apps/W1/EDocument/test/src/Receive/EDocManualImportTest.Codeunit.al:19]: Should we add a test that ensures that an error is thrown if there's a duplicate?

The tests for which message is shown is of low importance to me.

**Review by tinestaric — APPROVED** [review]
Looks good! Good work! :D

File: Apps/W1/EDocument/app/src/Integration/EDocIntegrationManagement.Codeunit.al (hunk 25 above)
Comment (petemchlk) [reply Apps/W1/EDocument/app/src/Integration/EDocIntegrationManagement.Codeunit.al:248]: Actually tried to avoid this commit, but happened to me while testing manually to get this error. Will keep this in my mind so if there will be some time left will try to handle this :)

File: Apps/W1/EDocument/test/src/Receive/EDocManualImportTest.Codeunit.al (hunk 27 above)
Comment (petemchlk) [reply Apps/W1/EDocument/test/src/Receive/EDocManualImportTest.Codeunit.al:19]: added test :)

**Review by tinestaric — DISMISSED** [review]

**Comment by nerkel91** [comment]
Implementation

With this update users can generate multiple incoming E-Documents when manually uploading multiple attachments from vendor documents.
Users can now manually upload multiple attachments from the **E-Documents** (ID 6122) page using the new **Import Files** action. The system will generate an individual E-Document for each uploaded file.
If E-Document already exists with the same:

- Incoming E-Document No

- Bill-to/Pay-to (Vendor No)

- Document Date

The system will generate a warning message to prevent duplication.

**Review by GMatuleviciute — DISMISSED** [review]
